### ISSUE TITLE
Feature: Synchronisation hooks

### ISSUE THREAD

**Issue by Ayplow** [browserpass/browserpass-native#80]
I want to keep the password store synchronised with a git repository, meaning my scripts need to keep track of changes in the local and remote copies. It seems like this use case could easily apply to things other than git, so I think a more generic approach is appropriate.

A `.git/hooks`-like structure seems fine:

```
~/.password-store/
  hooks/
    ...hook scripts
```

Off the top of my head, these would be handy;
### `start`

Run when the extension is started.

Also, a 'refresh' button would be added to the ui when this hook is detected, which would allow it to be run manually.

Once this hook completes, the cache should be refreshed

### `created`

Run when an entry is created with its' relative path

### `deleted`

Run when an entry is deleted with its' relative path

The hooks can be run with the store root as cwd

**Comment by erayd** [comment]
This feels like scope creep. Can you give me a sales pitch? I'm currently inclined to say no, but open to being convinced, and I can see there might be some utility in it. It's also not a breaking change, which is a positive.

If it happens, it probably belongs in `.browserpass.json`, rather than an entirely new folder.

Note that extension-start basically just means browser-start, but doesn't really have anything to do with the password store. Native app interaction occurs when browserpass is actually invoked (either via hotkey or by opening the popup).

/CC @maximbax

**Comment by Ayplow** [comment]
These are passwords to services designed to be able to be accessed from anywhere, so availability of our logins to them is important. For users who want to synchronise this information, it is part of their `pass` workflow. When using scripts, this addition can be invisible - its simply added to the same script that modifies your password store, while the user still enters one command to add/remove their passwords.

As I see it, the extension is fundamentally an interface for `pass` for ease of use in this context of online services. For UX, there is little difference between the [create password in external application > copy it to input] action that the extension has been designed to abstract and the current [create password in browser > copy it to external application] that is required for users in this situation.
_________________


I would argue that the difference between extension- and browser-start is relevant here - the goal is to keep the store 'ready' whenever the user is in need of it, and the user enabling the extension during a session signals their desire to use the store.

Using the json file sounds good - I admittedly just came across the project while looking for a better solution for password management, it was only a draft of how this could work. (json also avoids the problem of finding bat/sh/etc files depending on the platform)

Does the native app read from disk every time browserpass is invoked? I had assumed there was some layer of caching, which would've been the natural place to put that hook.

**Comment by erayd** [comment]
@Ayplow Thanks for that response :-).

> As I see it, the extension is fundamentally an interface for pass for ease of use in this context of online services.

It's not - the `pass` binary doesn't even need to be installed; the only host dependency is that `gpg` be present. Browserpass is a tool intended to use and interact with `pass`-style files on disk via `gpg`. The intention is to provide a tool that allows `pass` users to make use of the same password store within the context of a browser workflow.

> For users who want to synchronise this information, it is part of their pass workflow.

Yes. But trying to manage that synchronisation automatically is a significant hazard. I don't know whether you've has much to do with the Android password-store client, but one of the biggest pain points for them, over a very long time, has been trying to automate git synchronisation, which is something where people have so many different ways of working that it constantly runs into edge cases. Simply letting people take care of their own synchronisation removes a huge source of headaches.

With that said, hooks wouldn't do anything by default unless the user configured them, so that is a point in favour of adding the feature.

> I would argue that the difference between extension- and browser-start is relevant here...

Perhaps I was unclear earlier - I was pointing out that there _is no independent extension start event_ to which a hook could be attached. There is no point at which the extension 'starts up' that is distinct from the browser starting; it's a simultaneous action. There is no action required from the user to enable it; if the browser is running, then so is browserpass. We could run such a hook at browser start, but that seems fairly pointless noting many users leave their browser running more or less permanently, and very rarely cold-start it.

> Does the native app read from disk every time browserpass is invoked? I had assumed there was some layer of caching, which would've been the natural place to put that start hook.

There is no caching layer in browserpass. This is a deliberate design decision, that was made to avoid any possible situation in which browserpass might have a different view of the password store's state than what is actually present on disk.

The reason that this doesn't cause performance issues is because:
1. Browserpass doesn't decrypt anything during list operations, it only does a recursive glob of the password store, which is extremely fast even for many hundreds of entries.
2. The OS natively caches inode access, so even if this is on a spindle disk, it's still an extremely fast operation once the password store layout lands in the OS filesystem cache.

@maximbaz What do you think?

**Comment by erayd** [comment]
@maximbaz One other consideration - doing this opens up an attack surface that could allow an arbitrary code execution vulnerability on the host, in the event that somebody managed to compromise the browser extension. At the moment, that's not a possibility because the only thing the native component can run is `gpg`.

**Comment by Ayplow** [comment]
> allows `pass` users to make use of the same password store within the context of a browser workflow.

Certainly - all I meant by 'pass' here was the 'protocol' that [these programs](https://www.passwordstore.org/#other) work around.

>Simply letting people take care of their own synchronisation removes a huge source of headaches.

Yeah, completely automating any kind of workflow sounds pretty hard :) Though without hooks, it isnt even an option (okay, I guess you could watch the directory... but thats somewhat backwards)

> there is no independent extension start event

Seems like I was the one failing to communicate - all I meant was

`package.json`
```json
{
  "background": {
    "scripts": ["extension.js"]
  }
}
```

`extension.js`
```js
doStartHook()
if (startHookExists) createButton(doStartHook)
```

The reason I'd call that extension start rather than browser start is that the script is also run if someone enables the extension while the browser is running.

> doing this opens up an attack surface

Though the argument isnt invalid, you would also be arguing against every single rc/hook script that exists in linux. Generally, security is focused on never elevating the permissions that wouldve been needed to create a file, and otherwise assuming that hostile code doesnt get write access

**Comment by erayd** [comment]
> ...is also run if someone enables the extension while the browser is running.

That is so rare I don't think it's a case we should be caring about. In just about every single case, users leave an extension enabled, or they uninstall it. They don't generally turn it on and off the way you are implying.

That said, yes, background script startup is probably where a startup event should be triggered.

> ...you would also be arguing against every single rc/hook script that exists in linux.

None of the rest of those are exposed to code running in a web browser. They're isolated to the system, and require existing local code execution in order to mess with them. Putting such a hook in browserpass means there's potential exposure to web code if somebody finds an exploitable flaw in browserpass.

> Generally, security is focused on never elevating the permissions that wouldve been needed to create a file, and otherwise assuming that hostile code doesnt get write access.

It's also focused on sandboxing within the browser, and ensuring that hostile web code cannot get out. Defence in depth is a thing, and even if you're just running with user privileges, if you get out of the browser you can do an awful lot of damage.

**Comment by maximbaz** [comment]
Thanks for the idea @Ayplow! It's an interesting idea in principle, but I think it doesn't belong to this project (and browser in general), it rather deserves a separate tool of its own. 

As I understand, an example of what you want to achieve would be to run `git pull` on all your computers as soon as you create a password on one of them.

First, consider that users can create/delete password entries outside of browser (right now Browserpass doesn't have this functionality at all!), so it feels that a trigger should live outside of the browser as well.

The receiving end ideally would also not depend on a running browser, but rather be some sort of a daemon that starts on system startup and silently listens in background.

Remember also that you really want to run `git pull` not immediately after a password entry was _created_, but after it it was `git push`ed to a remote.

Finally, consider a UNIX philosophy, many small apps with single responsibilities are better than a single app that does it all 🙂 

----

If you or someone else would like to implement this tool, here are my recommendations:

1. The trigger could either be a git hook in your `.password-store` dir that is executed locally after `git push`, or if you host password store say on Github, there is a concept of "Webhooks" where Github sends a POST request on certain events (you'd go for `git push` in this case) - the latter approach will also support e.g. creating passwords on a mobile phone.

1. I'd probably go for something like [Firebase Cloud Firestore](https://firebase.google.com/docs/firestore/use-rest-api), so make your hook trigger a POST that adds an entry "created" or "deleted" to the db.

1. Then implement a tiny app that either listens to Firestore [using one of many available SDKs](https://firebase.google.com/docs/firestore/quickstart), or just queries the db via REST endpoint every now and then, and wrap it in a systemd service that runs on startup.

**Comment by Ayplow** [comment]
Oh I see! my mistake, I had seen the PR for generating passwords in the browser and assumed it was a feature of the extension. Yeah, this feature is much less useful if the extension doesn't yet make any modifications.

There is still the issue of being forces to drop to a cli if you're not sure if passwords have been updated, but I can understand that that is considered low priority.

Though a daemon started with the desktop session would *work*, I'm generally against keeping processes and sockets open when they're not necessary. Fetching that information when its relevant is much kinder to my already-suffering netbook.

As I will need this feature for my application, would you prefer I just fork the project and add it myself, or would it be better as an option in the extensions settings? Itll just be a path to a script to run when the browser starts, and will have a trigger in the UI.

**Comment by maximbaz** [comment]
Right now I don't think we will merge the functionality, so if you still prefer to implement it in the browser extension, I think go for a fork.

With Firebase I was hoping it would be really lightweight footprint on your system if you use their SDK, but yeah of course you could just go for a polling solution, or for that matter just a script that runs `git pull` every 10 minutes in your password store, and then you don't need to implement anything else 😉

**Comment by maximbaz** [comment]
Will close the ticket as out of scope, but do share the results of how you ended up implementing this, I'm sure other will find it interesting too!
